### PR TITLE
[inductor] Add ABI shim function for torch.scatter_reduce

### PR DESCRIPTION
**Pull request by ColinPeppler**
```
$ python test/inductor/test_aot_inductor.py -k test_scatter_reduce_fallback
----------------------------------------------------------------------
Ran 6 tests in 112.425s

OK (skipped=2)
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116424

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: test/inductor/test_aot_inductor.py [hunk 1]
```diff
@@ -1516,6 +1537,7 @@ def fail_minimal_arrayref_interface(is_skip=False):
     "test_repeat_output": fail_with_and_without_stack_allocation(is_skip=True),
     # the test segfaults
     "test_scatter_fallback": fail_with_and_without_stack_allocation(is_skip=True),
+    "test_scatter_reduce_fallback": fail_stack_allocation(is_skip=True),
```
Comment (ColinPeppler) [inline test/inductor/test_aot_inductor.py:1540]: use `fail_stack_allocation` or else you get this:

Error: cannot convert ‘torch::aot_inductor::ArrayRefTensor<long int>’ to ‘AtenTensorHandle’ {aka ‘AtenTensorOpaque*’}
  592 |     aoti_torch_scatter_reduce_out(buf0, buf0,0,arg1_1,arg2_1, "sum",1);

**Comment by ColinPeppler** [comment]
lint error seems to be unrelated?

```
$ proxy lintrunner -a
ok No lint issues.
```

File: test/inductor/test_aot_inductor.py (hunk 1 above)
Comment (aakhundov) [reply test/inductor/test_aot_inductor.py:1540]: cc @desertfire.

**Review by aakhundov — APPROVED** [review]
Thanks @ColinPeppler! Please fix the linter (by running `lintrunner -a`).

**Comment by aakhundov** [comment]
> lint error seems to be unrelated?

@ColinPeppler you're right, let me fix that one.

**Comment by aakhundov** [comment]
> > lint error seems to be unrelated?
> 
> @ColinPeppler you're right, let me fix that one.

Actually, seems that @malfet has already fixed that by reverting #115031. So just rebase.

**Comment by malfet** [comment]
In general, please base your PRs against `viable/strict` and you should never run into an issue when trunk is broken...

**Review by desertfire — APPROVED** [review]

**Pull request merged**